azkaban-common/src/main/java/azkaban/executor/container/watch/AzPodStatusDriver.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2021 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package azkaban.executor.container.watch;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import io.kubernetes.client.openapi.models.V1Pod;
+import io.kubernetes.client.util.Watch;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Drives execution {@link AzPodStatusListener} interface callbacks.
+ * Multiple listeners can be registered with a single instance of this class.
+ * Events received by this directly map the states in {@link AzPodStatus} and for every event all
+ * the callbacks registered for the particular event type are 'invoked'
+ *
+ * Invocation of the callbacks is done via pool of threads different from the one invoking this
+ * {@code onEvent} interface method of this class. This is done as a typical implementation of the
+ * {@link AzPodStatusListener} interface can include several blocking calls, for example for
+ * database status updates.
+ */
+@Singleton
+public class AzPodStatusDriver implements RawPodWatchEventListener {
+  private static final Logger logger = LoggerFactory.getLogger(AzPodStatusDriver.class);
+
+  private static final int THREAD_POOL_SIZE = 4;
+  public static final int SHUTDOW_TERMINATION_TIMEOUT_SECONDS = 5;
+  private final ExecutorService executor;
+  private final ImmutableMap<AzPodStatus, List<Consumer<AzPodStatusMetadata>>> listenerMap;
+
+  /**
+   * Create a new instance of {@link AzPodStatusDriver}.
+   */
+  @Inject
+  public AzPodStatusDriver() {
+    // Worth considering if we should get the thread-pool size from a config property instead.
```

Comment:
The thread-pool size must be configurable.

Reply:
Done.